## 高斯题目 

In [1]:
n = 100
Sum = 0
counter = 1
while counter<=n:
    Sum = Sum + counter
    counter += 1
    
print(Sum,counter)

5050 101


## 猜测密码本中的密钥

In [2]:
# __coding:utf-8__

name = 'password'
desk_path = '/Users/caimeijuan/github/myawesome/_src/'
full_path = desk_path + name + '.txt'
passwordFile = open(full_path)
secretPassword = passwordFile.read()
typedPassword = input("Enter your password:")

while typedPassword!=secretPassword :
    print('Access denied! PLease enter again.')
    typedPassword = input("Enter your password again:")
    if typedPassword == secretPassword:
        break    

if typedPassword == '123456':
    print('This is not a good password.')    
    print('Access granted.')
else:
    print('Access granted. Your password has %d characters.' % len(typedPassword)) 

Enter your password:5tgbvfr4
Access granted. Your password has 8 characters.


一直提示while语句语法错误，于是写一段while来看看，还怀疑是字符串判不同的问题。最后发现，`while`不小心大写成了`While`

## 验证用户名和密码 

In [3]:
while True:
    name = input('Who are you?')
    if name!='Joe':
        continue
    password = input('Hello, Joe. What is the password?(It\'s a fish.)')
    if password == 'swordfish':
        break
print('Access granted.')

Who are you?Joe
Hello, Joe. What is the password?(It's a fish.)swordfish
Access granted.


## 试图退出 

In [4]:
import sys
while True:
    response = input('Type exit to exit:')
    if response == 'exit':
        print('You typed "'+ response +'".')
        sys.exit()

Type exit to exit:exit
You typed "exit".


SystemExit: 

/Users/caimeijuan/anaconda/envs/python35/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3299: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.
  warn("To exit: use 'exit', 'quit', or Ctrl-D.", stacklevel=1)


## 月份随机数 

In [5]:
def getAnswer(answerNumber):
    print(answerNumber)
    if answerNumber == 1:
        return 'This is in January'
    elif answerNumber == 2:
        return 'This is in February'
    elif answerNumber == 3:
        return 'This is in March'
    elif answerNumber == 4:
        return 'This is in April'
    elif answerNumber == 5:
        return 'This is in May'
    elif answerNumber == 6:
        return 'This is in June'
    elif answerNumber == 7:
        return 'This is in July'
    elif answerNumber == 8:
        return 'This is in August'
    elif answerNumber == 9:
        return 'This is in September'
    elif answerNumber == 10:
        return 'This is in October'
    elif answerNumber == 11:
        return 'This is in November'
    elif answerNumber == 12:
        return 'This is in December'
    else:
        return 'This is not in any month of the year.'    

In [6]:
import random
getAnswer(random.randint(1,15))

13


'This is not in any month of the year.'

## print的参数

In [7]:
print("Hello","Ahcai","Pickle", sep=',',end="")
print("Wow!")

Hello,Ahcai,PickleWow!


## 异常语句

In [8]:
def spam(divideBy):
    try:
        return 42/divideBy
    except ZeroDivisionError:
        print("Error:Invalid argument.")    

In [9]:
print(spam(2))
print(spam(12))
print(spam(0))
print(spam(6))

21.0
3.5
Error:Invalid argument.
None
7.0


## 猜测100以内的随机数

In [10]:
import random
secret = random.randint(1,100)
times = 0
print("I\'m thinking of a number between 1 and 100. Please guess it:")

while True:
    guess = int(input())
    times +=1
    if guess < secret:
        print("Your guess is too low. Take a guess:")
    elif guess > secret:
        print("Your guess is too high. Take a guess:")
    else:
        print("Good job! Your guessed my number in %d times." %times) 
        break

I'm thinking of a number between 1 and 100. Please guess it:
50
Your guess is too high. Take a guess:
25
Your guess is too high. Take a guess:
12
Your guess is too high. Take a guess:
6
Your guess is too low. Take a guess:
9
Your guess is too low. Take a guess:
11
Good job! Your guessed my number in 6 times.


## Collatz序列

[冰雹猜想_百度百科](https://baike.baidu.com/item/%E5%86%B0%E9%9B%B9%E7%8C%9C%E6%83%B3/659469?fr=aladdin)

In [15]:
def collatz(number):
    if number%2 == 0:
        number = number//2
    else:
        number = number * 3 +1
    print(number)
    return number
    
number = int(input("Please input a positive integer："))
print('The initial number is: %d' %number)
print('The Follow-up digital sequence is:')

while True:
    if number!=1:
        number = collatz(number)
    else:
        break

Please input a positive integer：10
The initial number is: 10
The Follow-up digital sequence is:
5
16
8
4
2
1


## 逗号代码

In [17]:
def douhao(listPara):
    result = ""
    for i in range(len(listPara)-1):
        result = result + listPara[i] +','
    result = result[0:-1] +' and '+listPara[-1]
    return result
spam = ['ahcai', 'pickle', 'elephant', 'tiger', 'konglong']
print(douhao(spam))       

ahcai,pickle,elephant,tiger and konglong


## 字符图网格

In [18]:
grid = [['.','.','.','.','.','.'],['.','O','O','.','.','.'],['O','O','O','O','.','.'],['O','O','O','O','O','.'],['.','O','O','O','O','O'],['O','O','O','O','O','.'],['O','O','O','O','.','.'],['.','O','O','.','.','.'],['.','.','.','.','.','.'],]

In [19]:
print(grid)

[['.', '.', '.', '.', '.', '.'], ['.', 'O', 'O', '.', '.', '.'], ['O', 'O', 'O', 'O', '.', '.'], ['O', 'O', 'O', 'O', 'O', '.'], ['.', 'O', 'O', 'O', 'O', 'O'], ['O', 'O', 'O', 'O', 'O', '.'], ['O', 'O', 'O', 'O', '.', '.'], ['.', 'O', 'O', '.', '.', '.'], ['.', '.', '.', '.', '.', '.']]


In [24]:
for i in range(8):
    for j in range(5):
        print(grid[i][j],end=' ')
    print('\n')

. . . . . 

. O O . . 

O O O O . 

O O O O O 

. O O O O 

O O O O O 

O O O O . 

. O O . . 



## 储存生日信息

In [8]:
birthdays = {'Alice':'Apr 1', 'Bob':'Dec 12', 'Carol':'Mar 4'}
while True:
    print('Enter a name(blank to quit):')
    name = input()
    if name=="":
        break
    if name in birthdays:
        print(birthdays[name] + ' is the birthday of ' + name)
    else:
        print('I don\'t have birthday information for ' + name)
        bday = input('What is their birthday?')
        birthdays[name] = bday
        print('Birthday database updated.')

Enter a name(blank to quit):
Pickle
I don't have birthday information for Pickle
What is their birthday?July 28
Birthday database updated.
Enter a name(blank to quit):
Jimmy
I don't have birthday information for Jimmy
What is their birthday?Jan 30
Birthday database updated.
Enter a name(blank to quit):
Carol
Mar 4 is the birthday of Carol
Enter a name(blank to quit):
Jimmy
Jan 30 is the birthday of Jimmy
Enter a name(blank to quit):



In [9]:
birthdays

{'Alice': 'Apr 1',
 'Bob': 'Dec 12',
 'Carol': 'Mar 4',
 'Pickle': 'July 28',
 'Jimmy': 'Jan 30'}

## 数字符数

In [27]:
# __coding:utf8__

import pprint
def read_file(name):
    desk_path = '/Users/caimeijuan/github/myawesome/_src/'
    full_path = desk_path + name + '.txt'
    file = open(full_path,'r')
    return file

message = read_file("1").read()
count = {}
for character in message:
    count.setdefault(character,0)
    count[character] +=1

'''count = zip(count.values(),count.keys())
sorted(count,reverse=True) # 按值排序。但结果的数据类型发生了变化，不再是字典，而是zip格式。'''
# pprint.pprint(count) # 按键排序
print(pprint.pformat(count)) # 和上一句相比没有变化。

{' ': 1,
 '.': 3,
 '0': 1,
 '1': 2,
 '2': 1,
 '7': 1,
 '9': 1,
 '[': 1,
 ']': 1,
 '·': 2,
 '—': 2,
 '、': 1,
 '。': 1,
 '《': 1,
 '》': 1,
 '业': 1,
 '习': 1,
 '书': 1,
 '从': 1,
 '伯': 1,
 '何': 1,
 '克': 1,
 '出': 1,
 '利': 1,
 '到': 1,
 '刻': 1,
 '和': 1,
 '大': 1,
 '如': 1,
 '安': 1,
 '尔': 1,
 '工': 1,
 '师': 1,
 '平': 1,
 '年': 1,
 '序': 1,
 '德': 1,
 '志': 1,
 '意': 1,
 '手': 1,
 '推': 1,
 '斯': 1,
 '新': 1,
 '普': 1,
 '月': 1,
 '有': 1,
 '本': 1,
 '机': 1,
 '林': 1,
 '械': 1,
 '森': 1,
 '正': 1,
 '版': 2,
 '特': 1,
 '王': 1,
 '的': 1,
 '社': 1,
 '第': 1,
 '练': 1,
 '罗': 1,
 '美': 1,
 '艾': 1,
 '荐': 1,
 '著': 1,
 '译': 1,
 '阳': 1,
 '，': 3}


## 井字棋盘

In [61]:
theBoard = {'top-L':'','top-M':'','top-R':'','mid-L':'','mid-M':'','mid-R':'','low-L':'','low-M':'','low-R':''}
def printBoard(board):
    print('-+-+-+-')
    print('| '+ board['top-L'] + '| '+ board['top-M'] + '| '+ board['top-R'] + '| ')
    print('-+-+-+-')    
    print('| '+ board['mid-L'] + '| '+ board['mid-M'] + '| '+ board['mid-R'] + '| ')
    print('-+-+-+-')    
    print('| '+ board['low-L'] + '| '+ board['low-M'] + '| '+ board['low-R'] + '| ')
    print('-+-+-+-')    

board = theBoard.copy() # 直接赋值，会和列表一样，原棋盘也发生了变化。
turn = 'X'
for i in range(9):
    printBoard(board)
    move = input('Turn for ' + turn + '. Move on which space?')
    board[move] = turn
    if turn == 'X':
        turn = 'O'
    else:
        turn = 'X'
        
printBoard(board)

-+-+-+-
| | | | 
-+-+-+-
| | | | 
-+-+-+-
| | | | 
-+-+-+-
Turn for X. Move on which space?top-R
-+-+-+-
| | | X| 
-+-+-+-
| | | | 
-+-+-+-
| | | | 
-+-+-+-
Turn for O. Move on which space?low-M
-+-+-+-
| | | X| 
-+-+-+-
| | | | 
-+-+-+-
| | O| | 
-+-+-+-
Turn for X. Move on which space?mid-L
-+-+-+-
| | | X| 
-+-+-+-
| X| | | 
-+-+-+-
| | O| | 
-+-+-+-
Turn for O. Move on which space?top-M
-+-+-+-
| | O| X| 
-+-+-+-
| X| | | 
-+-+-+-
| | O| | 
-+-+-+-
Turn for X. Move on which space?top-L
-+-+-+-
| X| O| X| 
-+-+-+-
| X| | | 
-+-+-+-
| | O| | 
-+-+-+-
Turn for O. Move on which space?mid-M
-+-+-+-
| X| O| X| 
-+-+-+-
| X| O| | 
-+-+-+-
| | O| | 
-+-+-+-
Turn for X. Move on which space?low-R
-+-+-+-
| X| O| X| 
-+-+-+-
| X| O| | 
-+-+-+-
| | O| X| 
-+-+-+-
Turn for O. Move on which space?low-L
-+-+-+-
| X| O| X| 
-+-+-+-
| X| O| | 
-+-+-+-
| O| O| X| 
-+-+-+-
Turn for X. Move on which space?mid-R
-+-+-+-
| X| O| X| 
-+-+-+-
| X| O| X| 
-+-+-+-
| O| O| X| 
-+-+-+-


## 野餐物品清单

In [65]:
allGuests = {'Alice':{'apples':5, 'pretzels':12}, 'Cindy':{'ham sandwiches':5, 'cups':12}, 'Bob':{'ham sandwiches':3, 'apples':2}, 'Carol':{'cups':3, 'apple pies':1}}
def totalBrought(guests, item):
    numBrought = 0
    for k, v in guests.items():
        numBrought = numBrought + v.get(item,0)
    print('We have %d %s'%(numBrought,item))

print(allGuests)
totalBrought(allGuests,'apples')
totalBrought(allGuests,'pretzels')
totalBrought(allGuests,'ham sandwiches')
totalBrought(allGuests,'apple pies')
totalBrought(allGuests,'cups')


{'Alice': {'apples': 5, 'pretzels': 12}, 'Cindy': {'ham sandwiches': 5, 'cups': 12}, 'Bob': {'ham sandwiches': 3, 'apples': 2}, 'Carol': {'cups': 3, 'apple pies': 1}}
We have 7 apples
We have 12 pretzels
We have 8 ham sandwiches
We have 1 apple pies
We have 15 cups


## 画一个菱形

In [16]:
# __coding:utf-8__

def is_chinese(uchar):
    '''
    判断一个unicode是否是汉字
    '''
    if uchar >= u'\u4e00' and uchar<=u'\u9fa5':
        return True
    else:
        return False

def paintRhombusEnglish(n, strR): 
    '''
    该程序允许用户输入一个正奇数和一个英文字符或数字，画出边长为这个奇数的正方形中最大的菱形。
    '''
    for i in range(1,(n+1)//2):
        print((strR*(2*i-1)).center(n,'-')) # 如果不用'//'而是用'/'，则运算结果为浮点数，无法继续。
    for i in range((n+1)//2,0,-1):
        print((strR*(2*i-1)).center(n,'-'))

def paintRhombusChinese(n, strR): 
    '''
    该程序允许用户输入一个正奇数和一个汉字字符，画出边长为这个奇数的正方形中最大的菱形。/半个菱形。
    '''
    for i in range(1,(n+1)//2):
        print((strR*(2*i-1)).center(n+2*i,'-')) # 如果不用'//'而是用'/'，则运算结果为浮点数，无法继续。
    for i in range((n+1)//2,0,-1):
        print((strR*(2*i-1)).center(n+2*i,'-'))      
        
while True:
    '''
    如果用户输入的数字不是正奇数，则需要重新输入。
    '''
    n = input('Enter the length(for a odd number): ')
    if n.isdecimal() and int(n)>0 and int(n)%2==1:
        n = int(n)
        break

while True:
    '''
    如果用户输入的字符不是一位的字母或数字，则需要重新输入。
    '''
    strR = input('Enter the special string(only a number or alpha or Chinese character): ')
    if strR.isalnum() and len(strR)==1:
        break

'''
如果用户输入的字符是一个汉字，程序照旧运行，但是变形了。因为每个汉字占位2个字符。
如果用户输入了汉字，则调用汉字画法。如果用户输入了字母或数字，则调用英文画法。
'''
if is_chinese(strR)==False:
    paintRhombusEnglish(n, strR)
else:
    paintRhombusChinese(n, strR)

Enter the length(for a odd number): 9
Enter the special string(only a number or alpha or Chinese character): 福
-----福-----
-----福福福-----
-----福福福福福-----
-----福福福福福福福-----
-----福福福福福福福福福-----
-----福福福福福福福-----
-----福福福福福-----
-----福福福-----
-----福-----


## 打印整齐表格

In [1]:
def printPicnic(itemsDict, leftWidth, rightWidth):
    print('PICNIC ITEMS'.center(leftWidth+rightWidth, '-'))
    for k, v in itemsDict.items():
        print(k.ljust(leftWidth,'.')+str(v).rjust(rightWidth))
        
picnicItems = {'sandwiches':4, 'apples':12, 'cups':4, 'cookies':8000}
printPicnic(picnicItems,20, 6)

-------PICNIC ITEMS-------
sandwiches..........     4
apples..............    12
cups................     4
cookies.............  8000


## 密码箱

In [12]:
# python3
# pw.py
# __coding:utf-8__
# 这是一个密码箱

import sys
import pyperclip

PASSWORD = {"email":"eTfGIM2BZbYH7fxS", "blog":"a7QEHvM6Rszzo9Uj", "luggage":"12345"}

if len(sys.argv)<2:
    print("Usage: python pw.py [account] - copy account password")
    sys.exit()
    
account = sys.argv[1]

print(PASSWORD.get(account,"This account isn\'t exist."))

This account isn't exist.


## 合并列表为字典

In [1]:
list1 = ['key1','key2','key3']
list2 = ['1','2','3']
keyword = dict(zip(list1,list2))

In [2]:
type(keyword)

dict

In [3]:
print(keyword)

{'key1': '1', 'key2': '2', 'key3': '3'}


## 普通方法查找电话号码

In [28]:
def isPhoneNumber(text):
    if len(text)!= 12:
        return False
    for i in range(0,3):
        if not text[i].isdecimal():
            return False
    if text[3]!='-':
        return False        
    for i in range(4,7):
        if not text[i].isdecimal():
            return False
    if text[7]!='-':
        return False
    for i in range(8,12):
        if not text[i].isdecimal():
            return False
    return True

message = "Call me at 415-555-1011. 415-555-9999 is my office."

for i in range(len(message)):
    chunk = message[i:i+12]
    if isPhoneNumber(chunk):
        print('Phone number found: ' + chunk)
print('Done!')

Phone number found: 415-555-1011
Phone number found: 415-555-9999
Done!


## 用正则表达式查找电话号码

In [29]:
import re

phoneNumRegex = re.compile(r'\d{3}-\d{3}-\d{4}')
mo = phoneNumRegex.search(message)
print('Phone number found: ' + mo.group())

Phone number found: 415-555-1011


In [67]:
message1 = "Call me at (415)-(555-1011). (415)-(555-9999) is my office."

In [71]:
phoneNumRegex1 = re.compile(r'(\(\d{3}\))-(\(\d{3}-\d{4}\))')
mo = phoneNumRegex1.search(message1)
mo
'''
当message中没有括号时，又想用groups()，就需要在正则中加上()，message中不加。
如果message也有括号，正则中需要加上有转义符的括号，''里面的字符，依然要加上前一句说的()才能既匹配上message中的括号，又能使用groups()。
现在只返回第一个符合条件的。
'''

"\n当message中没有括号时，又想用groups()，就需要在正则中加上()，message中不加。\n如果message也有括号，正则中需要加上有转义符的括号，''里面的字符，依然要加上前一句说的()才能既匹配上message中的括号，又能使用groups()。\n现在只返回第一个符合条件的。\n"

##  查找电话号码和电子邮箱

In [94]:
import pyperclip
import re

# clipboard message
message = str(pyperclip.paste())
if message == "":
    print('Clipboard is null. Copy something to continue.')

# phone regex
phoneNumRegex = re.compile(r'''(
(\d{3}|\(\d{3}\))?             # area code
(\s|-|\.)?                     # seperator
(\d{3})                        # first 3 digits
(\s|-|\.)                      # seperator
(\d{4})                        # last 4 digits
(\s*(ext|x|ext.)\s*(\d{2,5}))? # extension
)''', re.VERBOSE)

# email regex
emailRegex = re. compile(r'''(
[a-zA-Z0-9._%+-]+              # username
@                              # @symbol
[a-zA-Z0-9._]+                 # domain name
(\.[a-zA-Z{2,4}])              # dot-something
)''', re.VERBOSE)

# Find all matches in clipboard text.
matches = []
for groups in phoneNumRegex.findall(message):
    phoneNum = '-'.join([groups[1],groups[3],groups[5]])
    if groups[8] != '':
        phoneNum += ' x' + groups[8]
    matches.append(phoneNum)
for groups in emailRegex.findall(message):
    matches.append(groups[0])

if len(matches) > 0:
    pyperclip.copy('\n'.join(matches))
    print('Done. To paste your results.')
    print('\n'.join(matches))
else:
    print('No phone numbers or emails found.')

No phone numbers or emails found.


## PI  

[Irrational Numbers Search Engine](http://www.subidiom.com/pi/pi.asp)

In [14]:
from math import pi
from decimal import *
import re
import numpy as np
import timeit

In [6]:
str(Decimal(pi))

'3.141592653589793115997963468544185161590576171875'

In [19]:
def pifomula(precision):
    getcontext().prec = precision
    fomula = 0
    for k in range(precision):
        fomula += 1/Decimal(16)**k * (Decimal(4)/(8*k+1)-Decimal(2)/(8*k+4)-Decimal(1)/(8*k+5)-Decimal(1)/(8*k+6))
    return fomula
t1 = timeit.Timer("pifomula(2*10**4)", "from __main__ import pifomula")
print("concat ",t1.timeit(number=2), "seconds")

concat  501.20280385 seconds


In [49]:
def findFromPI(numbers):
    strPI = str(pifomula(10))    
    if numbers.isdigit():
        if re.match(numbers, strPI) == None:
            print('Not found.')
        else:
            for match in re.finditer(numbers, strPI):
                print('The numeric string ' + numbers + ' appears at the ' + str(int(match.start())-1) + ' decimal digit of Pi.')
    else:
        print('Digitals please!')
findFromPI('14')

Not found.


In [44]:
print(str(pifomula(10)))

3.141592653


In [53]:
print(re.match('926', '3.141592653'))

None


## 从web抓取信息

### `webbrowser`

`webbrowser`这个模块，可以直接打开网址。

#### 打开单个城市

In [14]:
#_*_coding:utf-8_*_
import webbrowser
import re

def mapit(address):
    address = re.compile(r'\s+').sub('+',address) # 这里“+”之前不需要转义符\
    url = 'http://www.google.com/maps/place/' + address
    return webbrowser.open(url)

address = 'Wuxi,   Jiangsu,   China'    
mapit(address)

True

#### 批量打开城市群

In [17]:
import webbrowser

def mapcities(cities):
    address = []
    for city in cities:
        address.append(city + ', Jiangsu, China')
    for a in address:
        a = re.compile(r'\s+').sub('+',a)
        url = 'http://www.google.com/maps/place/' + a
        webbrowser.open(url)
        
cities = ['Wuxi', 'Suzhou', 'Xuzhou', 'Zhengjiang', 'Taizhou']
mapcities(cities)

### requests模块

In [42]:
#_*_coding:utf-8_*_
import requests
import chardet

In [39]:
url0 = 'http://www.gutenberg.org/cache/epub/1112/pg1112.txt'
url1 = 'http://example.webscraping.com'
url2 = 'http://www.engine3d.com'

In [40]:
res0 = requests.get(url0)
res1 = requests.get(url1)
res2 = requests.get(url2)

In [26]:
[type(res0),type(res1),type(res2)]

[requests.models.Response, requests.models.Response, requests.models.Response]

In [27]:
[res0.status_code == requests.codes.ok,res1.status_code == requests.codes.ok,res2.status_code == requests.codes.ok]

[True, True, True]

In [28]:
[len(res0.text),len(res1.text),len(res2.text)]

[178981, 7902, 17925]

In [43]:
print(res0.text[:250])

The Project Gutenberg EBook of Romeo and Juliet, by William Shakespeare

This eBook is for the use of anyone anywhere at no cost and with
almost no restrictions whatsoever.  You may copy it, give it away or
re-use it under the terms of the Proje


In [37]:
res2.encoding

'ISO-8859-1'

In [52]:
url3 = 'http://inventwithpython.com/page_that_does_not_exist'

In [53]:
res3 = requests.get(url3)

In [54]:
res3.raise_for_status()

HTTPError: 404 Client Error: Not Found for url: http://inventwithpython.com/page_that_does_not_exist

In [70]:
def download(url):
    res = requests.get(url)
    try:
        res.raise_for_status()
        html = res.text
        print('Webpage is download successfully.'+'\n'+'The beginning texts here:  '+'\n\n')
        print(html[:200])
    except Exception as exc :
        print('There was a problem: %s' %exc)

In [72]:
download(url0)

Webpage is download successfully.
The beginning texts here:  


The Project Gutenberg EBook of Romeo and Juliet, by William Shakespeare

This eBook is for the use of anyone anywhere at no cost and with
almost no restrictions whatsoever.  You may copy it, give 


In [71]:
download(url3)

There was a problem: 404 Client Error: Not Found for url: http://inventwithpython.com/page_that_does_not_exist


关于Unicode编码的知识：[The Absolute Minimum Every Software Developer Absolutely, Positively Must Know About Unicode and Character Sets (No Excuses!) – Joel on Software](https://www.joelonsoftware.com/2003/10/08/the-absolute-minimum-every-software-developer-absolutely-positively-must-know-about-unicode-and-character-sets-no-excuses/)

In [76]:
import requests

res = requests.get(url0)
res.raise_for_status()
playFile = open('RomeoAndJuliet.txt', 'wb')
for chunk in res.iter_content(100000):
    playFile.write(chunk)
playFile.close()

`encode_type = chardet.detect(html)`
`html = html.decode(encode_type['encoding'])`  

这里不是靠这两句解决问题的。已经测试过，html是str类型的。

用wb模式打开文件，写入内容是 `res.iter_content(100000)`  
作者说，使用iter_content是为了确保requests模块即使在**下载巨大**的文件时也**不会消耗太多**内存。

**The chunk size is the number of bytes it should read into memory. **

In [77]:
help(res.iter_content)

Help on method iter_content in module requests.models:

iter_content(chunk_size=1, decode_unicode=False) method of requests.models.Response instance
    Iterates over the response data.  When stream=True is set on the
    request, this avoids reading the content at once into memory for
    large responses.  The chunk size is the number of bytes it should
    read into memory.  This is not necessarily the length of each item
    returned as decoding can take place.
    
    chunk_size must be of type int or None. A value of None will
    function differently depending on the value of `stream`.
    stream=True will read data as it arrives in whatever size the
    chunks are received. If stream=False, data is returned as
    a single chunk.
    
    If decode_unicode is True, content will be decoded using the best
    available encoding based on the response.



requests文档地址：  
[Requests: HTTP for Humans™ — Requests 2.21.0 documentation](https://requests.readthedocs.io/en/master/)

### HTML

**苹果系统中，command+option+I，可以打开或关闭开发者工具，和Windows上的F12是一样的。**  

作者建议，**不要用正则表达式来解析HTML**。例如昨天遇到的将class写在a标签中间的那种，对于html来说仍然有效，用正则来预估所有的情况则会非常繁琐。专门用来解析html的模块，例如beautifulsoup，将更不容易出错。[html - RegEx match open tags except XHTML self-contained tags - Stack Overflow](https://stackoverflow.com/questions/1732348/regex-match-open-tags-except-xhtml-self-contained-tags/1732454#1732454)

### BeautifulSoup

In [81]:
import requests, bs4

url = 'http://nostarch.com'
res = requests.get(url)
res.raise_for_status()
noStarchSoup = bs4.BeautifulSoup(res.text)
type(noStarchSoup)

bs4.BeautifulSoup

In [92]:
exampleFile = open('UnicodeandCharacter.html')
exampleSoup = bs4.BeautifulSoup(exampleFile.read()) #.read()加了没加没区别。
type(exampleSoup)

bs4.BeautifulSoup

In [97]:
elms = exampleSoup.select('img')

In [117]:
homePage = 'https://www.joelonsoftware.com/'
url_src = homePage+elms[0]['src'][3:]
import webbrowser
webbrowser.open(url_src)

True

src="https://i1.wp.com/www.joelonsoftware.com/wp-content/uploads/2016/12/Pong.png?w=730&ssl=1"

为什么从网页中粘下来的src和上面elms的src不一样呢？

#### Google自动查询

In [145]:
# _*_coding:utf-8_*_
import requests, bs4, webbrowser,re

def googleit(query):
    
    #打开查询结果页面
    if '' in query:
        query = re.compile(r'\s+').sub('+', query)
    else:
        query = query

    url = 'http://www.google.com/search?q=' + query
    print('Googling...')
    res = requests.get(url)
    res.raise_for_status()
    
    #选择结果页面
    soup = bs4.BeautifulSoup(res.text)
    linkElems = soup.select('.r a')
    
    #打开前5个页面
    numOpen = min(5, len(linkElems))
    for i in range(numOpen):
        webbrowser.open('http://google.com' + linkElems[i].get('href'))
query = 'python webscraping'
googleit(query)

Googling...


In [177]:
#_*_coding:utf-8_*_
import requests, bs4,os

url = 'http://xkcd.com'
os.makedirs('xkcd', exist_ok=True)
while not url.endswith('#'):
    #下载网页
    print('Downloading page %s ...'%url)
    res = requests.get(url)
    res.raise_for_status()
    soup = bs4.BeautifulSoup(res.text)
    
    #寻找漫画
    comicElem = soup.select('#comic img')
    if comicElem == []:
        print('Could not find comic image.')
    else:
        try:
            comicUrl = 'http:'+ comicElem[0].get('src')
            print('Downloading image is %s ...'%comicUrl)
            res = requests.get(comicUrl)
            res.raise_for_status()  
        except:
            comicUrl = 'http://www.xkcd.com'+ comicElem[0].get('src')
            print('That is wrong. Downloading image is %s ...'%comicUrl)    
            res = requests.get(comicUrl)
            res.raise_for_status() 

    #保存漫画
    imageFile = open(os.path.join('xkcd', os.path.basename(comicUrl)), 'wb')
    for chunk in res.iter_content(100000):
        imageFile.write(chunk)
    imageFile.close() 

    #找前一张漫画
    prevLink = soup.select('a[rel="prev"]')[0]
    url = 'http://xkcd.com' + prevLink.get('href')
    
print('Done.')

That is wrong. Downloading image is http://www.xkcd.com/2067/asset/challengers_header.png ...


SSLError: HTTPSConnectionPool(host='xkcd.com', port=443): Max retries exceeded with url: /1813/ (Caused by SSLError(SSLError("bad handshake: SysCallError(-1, 'Unexpected EOF')")))

In [176]:
url = 'https://xkcd.com/2031/'
res = requests.get(url)
res.raise_for_status()
soup = bs4.BeautifulSoup(res.text)
comicElem = soup.select('#comic img')
try:
    comicUrl = 'http:'+ comicElem[0].get('src')
    print('Downloading image is %s ...'%comicUrl)
    res = requests.get(comicUrl)
    res.raise_for_status()  
except:
    comicUrl = 'http://www.xkcd.com'+ comicElem[0].get('src')
    print('That is wrong. Downloading image is %s ...'%comicUrl)    
    res = requests.get(comicUrl)
    res.raise_for_status() 

明白了，这里用的是canvas，不是img。

2031和1813都出现的错误：  

SysCallError                              Traceback (most recent call last)

SSLError: HTTPSConnectionPool(host='xkcd.com', port=443): Max retries exceeded with url: /1813/ (Caused by SSLError(SSLError("bad handshake: SysCallError(-1, 'Unexpected EOF')")))

### selenium

In [24]:
from selenium import webdriver
import os

In [35]:
path = '/Users/caimeijuan/anaconda/envs/python35/lib/python3.7/site-packages/selenium/webdriver/chrome/chromedriver'
browser = webdriver.Chrome(path)

折腾过程见教程。

In [36]:
type(browser)

selenium.webdriver.chrome.webdriver.WebDriver